Server Sent Events example - issue #7008

ewaldc opened the pull request:
Illustrates the use of SSE using ESP8266WebServer

*edit from maintainer:* closes #7008

d-a-v commented:
CI has this error (on `continuous-integration-...detail` button below)
```
/home/travis/build/esp8266/Arduino/libraries/ESP8266WebServer/examples/ServerSentEventsMultiClient/ServerSentEventsMultiClient.ino:161:22: error: comparison is always true due to limited range of data type [-Werror=type-limits]
       if (channel >= 0)
                      ^
cc1plus: all warnings being treated as errors
```

Style check can be solved by the script in `tests/restyle.sh` (then commit the differences)

d-a-v commented:
You also need, in this PR, on your local repository, to remove the submodule update you added by error:

Probably with
```
git checkout upstream/master -- libraries/LittleFS/lib/littlefs
```
and `libraries/SoftwareSerial`, `tools/sdk/ssl/bearssl`, `tools/esptool` and `tools/sdk/lwip2/builder` then commit push as usual

ewaldc commented:
Done, apologies for my mistake and for wasting developer cycles, there is so much I have to learn about GitHub...  
Hopefully I did it right.

d-a-v commented:
Did you push ?

ewaldc commented:
Yes, https://github.com/ewaldc/Arduino is fully up to date and equal to my local repository.
The example, which is really the only thing I'd like to contribute, is [here](https://github.com/ewaldc/Arduino/tree/master/libraries/ESP8266WebServer/examples/ServerSentEvents )

Martin-Laclaustra commented:
@ewaldc 
Thank you. This was an inspiring example.
I found a couple of corrections for your code:
https://github.com/ewaldc/Arduino/blob/81ec6e8ae6a68d55e0435c3212dbd419cc9144f5/libraries/ESP8266WebServer/examples/ServerSentEvents/ServerSentEvents.ino#L86
to:
`      Serial.printf_P(PSTR("SSEKeepAlive - client is still listening on channel %d\n"), i);`
https://github.com/ewaldc/Arduino/blob/81ec6e8ae6a68d55e0435c3212dbd419cc9144f5/libraries/ESP8266WebServer/examples/ServerSentEvents/ServerSentEvents.ino#L89
to:
`      Serial.printf_P(PSTR("SSEKeepAlive - client not listening on channel %d, remove subscription\n"), i);`
(the variable for the printf was missing)
and
https://github.com/ewaldc/Arduino/blob/81ec6e8ae6a68d55e0435c3212dbd419cc9144f5/libraries/ESP8266WebServer/examples/ServerSentEvents/ServerSentEvents.ino#L146
to
`  if (val != newVal) SSEBroadcastState(name, val, newVal); // only broadcast if state is different`
(values are interchanged)

Martin-Laclaustra commented:
@ewaldc
I thought that I could add a javascript client side to the example to make it self-contained.
The html page is the attached file (as a header, to be served through the same script).
[webpages.h.zip](https://github.com/esp8266/Arduino/files/4144157/webpages.h.zip)
And in the sketch, only two lines have to be added:
After line 41:
`#include "webpages.h"`
After:
https://github.com/ewaldc/Arduino/blob/81ec6e8ae6a68d55e0435c3212dbd419cc9144f5/libraries/ESP8266WebServer/examples/ServerSentEvents/ServerSentEvents.ino#L174
`  server.on("/", HTTP_GET, [](){server.send(200, "text/html", FPSTR(webpagesseclient));});`
I hope you find this convenient.

Martin-Laclaustra commented:
@ewaldc 
I have a couple of extra suggestions:
- Dealing with the channels with specific callbacks instead of using `server.onNotFound(handleAll);`. That way it allows coupling this example with other examples serving content from the filesystem. It also makes the sketch shorter. I already worked on this and have functioning code. Do not hesitate to ask me and I will share this with you.
- Sending data without "event: event", so it can be received with the javascript `.onmessage`, which is the standard example you find online for using SSE. I also have this modification (only a couple of lines, but in both files .ino and .h).
- Fix `subscriptionCount` to reflect actual used channels. Currently it never decreases.
- Releasing subscriptions if they are not connected after some timeout (currently this only happens if there is at least one active connection and it can lead to running out of channels without possiblility of release).
- Creating a simpler example without channels (only one connection per IP), so that the server works in a way compatible with most online javascript examples.
(I am working in the later points).
Best

ewaldc commented:
@Martin-Laclaustra, 
thanks for the corrections, I have posted an **[update](https://github.com/ewaldc/Arduino/blob/master/libraries/ESP8266WebServer/examples/ServerSentEvents/ServerSentEvents.ino)**.   The example is a simplified version of a more complex code posted [_here_](https://github.com/ewaldc/OpenHAB-ESP), which also works with HTML5 event listeners in Javascript and [JAVA SSE from here maps](https://github.com/heremaps/oksse). 
On your suggestions:

- _Dealing with the channels with specific callbacks instead of using server.onNotFound(handleAll);._ 
I could not make that work.  The problem is that I did not find a call to unregister a callback: no way to invalidate a previous call to _server.on()_  once a channel is invalid.  In de real code, a channel is a unique UUID, and I ended up with hundreds of active callbacks on invalid UUID's, all consuming 
memory and slowing down processing of incoming web requests.  I also started to use server.on() in a sparse way in general due to its use of strings and extensive memory footprint.
But maybe you have a solution...

- _Sending data without "event: event", so it can be received with the javascript .onmessage, which is the standard example you find online for using SSE_
The [EventSource standard](https://html.spec.whatwg.org/multipage/server-sent-events.html) recommends the use of messages to separate between events (of different types).   **[Javascript EventSource](https://javascript.info/server-sent-events)**  supports _data_, _id's_ and _events_, so I fail to see the issue with javascript...   That said, you can of course, simplify the code to use events containing only _data_ .   One other reason why I decided the keep it was that initially I struggled with the keepAlive timer crashing my code.  It turned out that I used the wrong Ticker call, hence I decided to add that in so others won't spend hours debugging multi-thread issues...

- _Fix subscriptionCount to reflect actual used channels. Currently it never decreases._
Oversight in my side, it does of course in my full code...  Fixed

- _Creating a simpler example without channels (only one connection per IP), so that the server works in a way compatible with most online javascript examples_.
Originally I posted two examples., with exactly those differences: only one registration per IP address and no channels.  But as I was implementing changes to fix certain issues (e.g. WDT reset, nodelay causes crash etc. I decided to go back to one (the more elaborate one) as it was getting complex to update the GitHub PR's.   That said, I still have both examples !

Originally I added the webclient, but I felt that I should then also post a Javascript example to make it all work, which in turn would not fit into one single .ino file...  Hence I settled for just using curl and adding the HOW TO's just in comments.  Maybe I took the lazy way out here :-)

Hope it helps.
Ewald

Martin-Laclaustra commented:
> The example is a simplified version of a more complex code posted [_here_](https://github.com/ewaldc/OpenHAB-ESP), which also works with HTML5 event listeners in Javascript and [JAVA SSE from here maps](https://github.com/heremaps/oksse).

Thank you. I will have a look.

>     *  _Dealing with the channels with specific callbacks instead of using server.onNotFound(handleAll);._
>       I could not make that work.  The problem is that I did not find a call to unregister a callback: no way to invalidate a previous call to _server.on()_  once a channel is invalid.  In de real code, a channel is a unique UUID, and I ended up with hundreds of active callbacks on invalid UUID's, all consuming
>       memory and slowing down processing of incoming web requests.  I also started to use server.on() in a sparse way in general due to its use of strings and extensive memory footprint.
>       But maybe you have a solution...

It is true that you can not unregister in the ESP server, but that is not necessary. My solution pre-registers the channels (I am assuming they are a reasonable number). _"Hundreds of active callbacks"_  Do you expect hundreds of concurrent channels handled by the ESP? My current example works with single figure channels but it can be easily extended to handle two figure channels.
Here there is a patch against your last update:
[channel_handles.patch.zip](https://github.com/esp8266/Arduino/files/4157691/channel_handles.patch.zip)

> Originally I added the webclient, but I felt that I should then also post a Javascript example to make it all work, which in turn would not fit into one single .ino file... Hence I settled for just using curl and adding the HOW TO's just in comments. Maybe I took the lazy way out here :-)

Why do you not consider the one that I uploaded in my previous post? It is kind of a minimal-size example in a single file. You could paste it instead of the include (although I think it is clearer as a separate header file). It is only 58 lines. I can even be further shortened if some unnecessary html or javascript code is removed from it.

Thank you for your attention.

ewaldc commented:
Subscription channels are rarely implemented with just a simple digit as in the example.  For example in Eclipse Smarthome, OpenHab and several other IOT frameworks, a (unique) UUID is being used. Security is an important factor in their decision to do so (e.g. avoid unwanted listeners).  This invalides pre-registration based solutions. 

In addition, some clients register an event listener for each page of their GUI (to listen for object updates) and unregister after the user leaves that page (e.g. OpenHab Android/IOS client).  So yes, in a typical day my application handles literally hundreds of subscriptions, some days closer to a thousand.  Besides not being able to unregister (causing the ESP to run out of memory), the ESP web server will execute a string compare for every single registered subscription. When dealing with e.g. long paths (not almost in my control) ("http://myiotserver.myverylongdomainname.net:48888/rest/events/listener/<UUID>")  we found out (the hard way) that using _server.on()_ for each subscription consumed not only a lot of RAM (e.g. due the use of the String class) but can slow down the web server to unuseable levels, including hitting soft WDT resets.  Hence, why I reverted back to doing _one_ single string comparison to find out whether we are dealing with an event callback or not. 

In fact, I ended almost not using _server.on()_ at all saving lots of memory.  If you have hundreds of IOT objects, each of which can be queried by "<URI>/rest/item/<name>" , this issue of excessive _server.on()_ calls/string comparisons/memory use only gets worse.

If the ESP Arduino library would handle URL patterns, or match the beginning of the string e.g. _server.onStartsWith()_ that would be a whole different case of course.   I actually made a PR to implement such a function, plus the possibility to unregister, made modifications to the String class to avoid excessive copying etc but before I knew, it had become a massive PR which was difficult to explain, unless I reverted back to posting 20 or so individual issues, each with their own PR and code change. Rightfully so, the maintainers keep a close eye on compatibility and memory footprint, so it's not obvious to defend the addition of new functions.

Hopefully, this gives some insights in why things were done differently, even though you have a very valid point that there are simpler and "less-lines-of-code" ways to make the example work.
But I am also a believer in posting examples that trade off some simplicity for real world deployment experience, scalability and performance, specifically in the embedded IOT world.

devyte commented on libraries/ESP8266WebServer/examples/ServerSentEvents/ServerSentEvents.ino:
```diff
@@ -0,0 +1,218 @@
+  if (strncmp_P(uri, PSTR("/rest/events/"), sizeof("/rest/events"))) {
+    return handleNotFound();
+  }
+  uri += sizeof("/rest/events");
```
isn't there a missing / here? I'm guessing the uri should be "rest/events/<channel>"

earlephilhower replied:
No, thanks to C strings having the \0.  `sizeof("abc") == 4` since there's the terminal NUL.  This uses the \0 to skip the '/'.  The script does seem to work well in my own local testing.

devyte commented on libraries/ESP8266WebServer/examples/ServerSentEvents/ServerSentEvents.ino:
```diff
@@ -0,0 +1,218 @@
+void SSEHandler(uint8_t channel) {
+  WiFiClient client = server.client();
+  SSESubscription &s = subscription[channel];
+  if (s.clientIP != uint32_t(client.remoteIP())) { // IP addresses don't match, reject this client
```
why cast to uint32_t?

earlephilhower commented on libraries/ESP8266WebServer/examples/ServerSentEvents/ServerSentEvents.ino:
```diff
@@ -0,0 +1,218 @@
+
+void handleAll() {
+  const char *uri = server.uri().c_str();
+  if (strncmp_P(uri, PSTR("/rest/events/"), sizeof("/rest/events"))) {
```
Actually, it's used.  Same thing as just below, sizeof(c-string) includes the NUL, so he's strncmp(uri, "/rest/events/", strlen("/rest/events")).

I'll rewrite it that way for sanity, even though what's there now is technically correct (but confusing).

earlephilhower replied:
I have swapped this for a more typical `strlen()` based offset/comparison block.  The original code did work fine, but I fear was a little too smart for the typical Arduino crowd.  Don't want them thinking you can do `char *str="lala"; int strLen = sizeof(str);`  or to confuse them about the terminal \0 being part of sizeof() but not strlen().

ewaldc replied:
Sorry, always trying to squeeze out the extra nano-second :-)
Thanks for accepting and improving this example.

earlephilhower commented on libraries/ESP8266WebServer/examples/ServerSentEvents/ServerSentEvents.ino:
```diff
@@ -0,0 +1,218 @@
+
+void handleSubscribe() {
+  uint8_t channel;
+  IPAddress clientIP = server.client().remoteIP();   // get IP address of client
```
Can't quite do it that way.  The way it works is you connect and get a channel to listen to.  You then open a new connection to that channel later on.  So, the `WiFiClient` you handle subscribe from will *never* be the `WiFiClient` you'll actually be sending events to.

And, even in the event notifier, the `WiFiClient` may change after the first connection.  The channel is still good even when you have disconnections/network hiccups, so you may have multiple connections (at different times) from the same IP...

earlephilhower commented on libraries/ESP8266WebServer/examples/ServerSentEvents/ServerSentEvents.ino:
```diff
@@ -0,0 +1,218 @@
+    if (!(subscription[i].clientIP)) {
+      continue;
+    }
+    WiFiClient client = subscription[i].client;
```
Done

earlephilhower commented on libraries/ESP8266WebServer/examples/ServerSentEvents/ServerSentEvents.ino:
```diff
@@ -0,0 +1,218 @@
+}
+
+// Simulate sensors
+void updateSensor(const char* name, unsigned short *value) {
```
Done, and reworked to take a struct to fix a bug.

earlephilhower commented on libraries/ESP8266WebServer/examples/ServerSentEvents/ServerSentEvents.ino:
```diff
@@ -0,0 +1,218 @@
+  size_t offset = SSEurl.length();
+  SSEurl += F("/rest/events/");
+
+  if (subscriptionCount == SSE_MAX_CHANNELS - 1) {
```
Done

earlephilhower commented on libraries/ESP8266WebServer/examples/ServerSentEvents/ServerSentEvents.ino:
```diff
@@ -0,0 +1,218 @@
+
+#define SSE_MAX_CHANNELS 8  // in this simplified example, only eight SSE clients subscription allowed
+struct SSESubscription {
+  uint32_t clientIP;
```
Done
